### PR TITLE
Add misc derivations

### PR DESCRIPTION
**Pull request by realmfoo**
Added derive support with CKM_CONCATENATE_DATA_AND_BASE, CKM_CONCATENATE_BASE_AND_DATA and CKM_CONCATENATE_BASE_AND_KEY.

### CODE REVIEW

File: src/lib/SoftHSM.cpp [hunk 1]
```diff
@@ -11383,23 +11524,76 @@ CK_RV SoftHSM::deriveSymmetric
...
+								  otherKey->getBooleanValue(CKA_SENSITIVE, true);
+				if (bSensitive) {
+					bOK = bOK && osobject->setAttribute(CKA_SENSITIVE, true);
+				}
+				// If either of the two original keys has its CKA_EXTRACTABLE attribute
+				// set to CK_FALSE, so does the derived key.  If not, then the derived
+				// key’s CKA_EXTRACTABLE attribute is set either from the supplied template
+				// or from a default value.
+				bool bExtractable = baseKey->getBooleanValue(CKA_EXTRACTABLE, false) ||
+									otherKey->getBooleanValue(CKA_EXTRACTABLE, false);
```
Comment (Aearsis) [inline src/lib/SoftHSM.cpp]: The logic here is wrong, IMHO. One of the keys being non-extractable should make the resulting key non-extractable.

```suggestion
				bool bExtractable = baseKey->getBooleanValue(CKA_EXTRACTABLE, false) &&
									otherKey->getBooleanValue(CKA_EXTRACTABLE, false);
```

File: src/lib/SoftHSM.cpp [hunk 2]
```diff
@@ -11489,23 +11524,76 @@ CK_RV SoftHSM::deriveSymmetric
...
+				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bAlwaysSensitive);
+				// The derived key’s CKA_NEVER_EXTRACTABLE attribute is set to CK_TRUE
+				// if and only if both of the original keys have their CKA_NEVER_EXTRACTABLE
+				// attributes set to CK_TRUE
+				bool bNeverExtractable = baseKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, false) &&
+										 otherKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, false);
+				bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bNeverExtractable);
+			}
+			else if (pMechanism->mechanism == CKM_CONCATENATE_BASE_AND_DATA ||
+				 pMechanism->mechanism == CKM_CONCATENATE_DATA_AND_BASE)
```
Comment (Aearsis) [inline src/lib/SoftHSM.cpp:11561]: Probably just an irrelevant nitpick, if this PR is accepted as a whole: this branch is added only as a part of the commit introducing `CKM_CONCATENATE_BASE_AND_KEY`.

File: src/lib/SoftHSM.cpp [hunk 3]
```diff
@@ -11383,23 +11524,76 @@ CK_RV SoftHSM::deriveSymmetric
...
+					bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, bAlwaysSensitive);
+				} else {
+					bOK = bOK && osobject->setAttribute(CKA_ALWAYS_SENSITIVE, false);
+				}
+				if (baseKey->getBooleanValue(CKA_NEVER_EXTRACTABLE, true)) {
+					bool bNeverExtractable = osobject->getBooleanValue(CKA_EXTRACTABLE, false) == false;
+					bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE, bNeverExtractable);
+				} else {
+					bOK = bOK && osobject->setAttribute(CKA_NEVER_EXTRACTABLE, false);
+				}
```
Comment (Aearsis) [inline src/lib/SoftHSM.cpp:11596]: What cases does this else branch cover?

File: src/lib/test/DeriveTests.cpp [hunk 4]
```diff
@@ -864,3 +890,153 @@ void DeriveTests::testSymDerive()
...
+    CK_OBJECT_HANDLE hDerive = CK_INVALID_HANDLE;
+
+    CK_BYTE data[] = { 0x01, 0x23, 0x45, 0x67 };
+    CK_KEY_DERIVATION_STRING_DATA param = { &data[0], sizeof(data) };
+    CK_MECHANISM mechanism = { CKM_CONCATENATE_DATA_AND_BASE, &param, sizeof(param)};
+
+    CK_ULONG bytes = 16;
+    CK_BBOOL bTrue = CK_TRUE;
+    CK_ATTRIBUTE keyAttribs[] = {
+            { CKA_EXTRACTABLE, &bTrue, sizeof(bTrue) },
```
Comment (Aearsis) [inline src/lib/test/DeriveTests.cpp:963]: I would like to see a testcase for `CKA_EXTRACTABLE` being false, as it's quite important to get it right :) I can prepare it if you want, just ask for it.

**Review by Aearsis — CHANGES_REQUESTED** [review]

File: src/lib/SoftHSM.cpp (hunk 1 above)
Comment (realmfoo) [reply src/lib/SoftHSM.cpp]: You're correct. I've fixed it and added test cases.

File: src/lib/SoftHSM.cpp (hunk 2 above)
Comment (realmfoo) [reply src/lib/SoftHSM.cpp:11561]: Should I split this commit?

File: src/lib/SoftHSM.cpp (hunk 3 above)
Comment (realmfoo) [reply src/lib/SoftHSM.cpp:11596]: It's a default branch for other mechanisms.

File: src/lib/test/DeriveTests.cpp (hunk 4 above)
Comment (realmfoo) [reply src/lib/test/DeriveTests.cpp:963]: I've added testcases for `CKA_EXTRACTABLE` and `CKA_SENSITIVE`.

File: src/lib/SoftHSM.cpp (hunk 2 above)
Comment (Aearsis) [reply src/lib/SoftHSM.cpp:11561]: Certainly not because of me.

File: src/lib/SoftHSM.cpp (hunk 3 above)
Comment (Aearsis) [reply src/lib/SoftHSM.cpp:11596]: Sorry, did not notice that this code is just moved into else branch. I thought it was added, which seemed suspicious. OK then.

**Comment by realmfoo** [comment]
@Aearsis please let me know if you need any additional actions from my end...

**Comment by Aearsis** [comment]
> @Aearsis please let me know if you need any additional actions from my end...

Don't be mistaken, I'm not a maintainer :) This is more for @halderen or @rijswijk .

**Comment by rijswijk** [comment]
Thanks for the contribution, I reviewed the code, it looks good to me. Asking @halderen if he concurs, and if so, to merge it into `develop`

**Comment by halderen** [comment]
Reviewed and contains the checks regarding the sensitive and extractable attributes, that are absolutely needed.

**Pull request merged**